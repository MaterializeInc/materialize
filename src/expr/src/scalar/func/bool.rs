// Copyright Materialize, Inc. All rights reserved.
//
// Use of this software is governed by the Business Source License
// included in the LICENSE file.
//
// As of the Change Date specified in that file, in accordance with
// the Business Source License, use of this software will be governed
// by the Apache License, Version 2.0.

//! Boolean functions.

use repr::strconv;
use repr::{Datum, RowArena, ScalarType};

use crate::scalar::func::{FuncProps, Nulls, OutputType};
use crate::scalar::{EvalError, ScalarExpr};

pub const CAST_BOOL_TO_STRING_PROPS: FuncProps = FuncProps {
    can_error: false,
    preserves_uniqueness: true,
    nulls: Nulls::Sometimes {
        propagates_nulls: true,
        introduces_nulls: false,
    },
    output_type: OutputType::Fixed(ScalarType::String),
};

pub fn cast_bool_to_string_explicit(a: Datum) -> Result<Datum, EvalError> {
    // N.B. this function differs from `cast_bool_to_string_implicit` because
    // the SQL specification requires `true` and `false` to be spelled out
    // in explicit casts, while PostgreSQL prefers its more concise `t` and `f`
    // representation in implicit casts.
    match a.unwrap_bool() {
        true => Ok(Datum::from("true")),
        false => Ok(Datum::from("false")),
    }
}

pub fn cast_bool_to_string_implicit(a: Datum) -> Result<Datum, EvalError> {
    Ok(Datum::String(strconv::format_bool_static(a.unwrap_bool())))
}

pub const CAST_STRING_TO_BOOL_PROPS: FuncProps = FuncProps {
    can_error: true,
    preserves_uniqueness: false,
    nulls: Nulls::Sometimes {
        propagates_nulls: true,
        introduces_nulls: false,
    },
    output_type: OutputType::Fixed(ScalarType::Bool),
};

pub fn cast_string_to_bool(a: Datum) -> Result<Datum, EvalError> {
    match strconv::parse_bool(a.unwrap_str())? {
        true => Ok(Datum::True),
        false => Ok(Datum::False),
    }
}

pub const AND_PROPS: FuncProps = FuncProps {
    can_error: false,
    preserves_uniqueness: false,
    nulls: Nulls::Sometimes {
        propagates_nulls: false,
        introduces_nulls: false,
    },
    output_type: OutputType::Fixed(ScalarType::Bool),
};

pub fn and<'a>(
    datums: &[Datum<'a>],
    temp_storage: &'a RowArena,
    a_expr: &'a ScalarExpr,
    b_expr: &'a ScalarExpr,
) -> Result<Datum<'a>, EvalError> {
    match a_expr.eval(datums, temp_storage)? {
        Datum::False => Ok(Datum::False),
        a => match (a, b_expr.eval(datums, temp_storage)?) {
            (_, Datum::False) => Ok(Datum::False),
            (Datum::Null, _) | (_, Datum::Null) => Ok(Datum::Null),
            (Datum::True, Datum::True) => Ok(Datum::True),
            _ => unreachable!(),
        },
    }
}

pub const OR_PROPS: FuncProps = FuncProps {
    can_error: false,
    preserves_uniqueness: false,
    nulls: Nulls::Sometimes {
        propagates_nulls: false,
        introduces_nulls: false,
    },
    output_type: OutputType::Fixed(ScalarType::Bool),
};

pub fn or<'a>(
    datums: &[Datum<'a>],
    temp_storage: &'a RowArena,
    a_expr: &'a ScalarExpr,
    b_expr: &'a ScalarExpr,
) -> Result<Datum<'a>, EvalError> {
    match a_expr.eval(datums, temp_storage)? {
        Datum::True => Ok(Datum::True),
        a => match (a, b_expr.eval(datums, temp_storage)?) {
            (_, Datum::True) => Ok(Datum::True),
            (Datum::Null, _) | (_, Datum::Null) => Ok(Datum::Null),
            (Datum::False, Datum::False) => Ok(Datum::False),
            _ => unreachable!(),
        },
    }
}

pub const NOT_PROPS: FuncProps = FuncProps {
    can_error: false,
    preserves_uniqueness: true,
    nulls: Nulls::Sometimes {
        propagates_nulls: true,
        introduces_nulls: false,
    },
    output_type: OutputType::Fixed(ScalarType::Bool),
};

pub fn not(a: Datum) -> Result<Datum, EvalError> {
    Ok(Datum::from(!a.unwrap_bool()))
}
