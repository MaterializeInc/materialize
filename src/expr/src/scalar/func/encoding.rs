// Copyright Materialize, Inc. All rights reserved.
//
// Use of this software is governed by the Business Source License
// included in the LICENSE file.
//
// As of the Change Date specified in that file, in accordance with
// the Business Source License, use of this software will be governed
// by the Apache License, Version 2.0.

//! Encoding and decoding support for various formats that represent binary data
//! as text data.

use ore::ascii::UncasedStr;
use ore::fmt::FormatBuffer;
use repr::strconv;

use crate::EvalError;

/// An encoding format.
pub trait Format {
    /// Encodes a byte slice into its string representation according to this
    /// format.
    fn encode(&self, bytes: &[u8]) -> String;

    /// Decodes a byte slice from its string representation according to this
    /// format.
    fn decode(&self, s: &str) -> Result<Vec<u8>, EvalError>;
}

/// PostgreSQL-style Base64 encoding.
///
/// PostgreSQL follows RFC 2045, which requires that lines are broken after 76
/// characters when encoding and that all whitespace characters are ignored when
/// decoding. See <http://materialize.com/docs/sql/functions/encode> for
/// details.
struct Base64Format;

impl Base64Format {
    const CHARSET: &'static [u8] =
        b"ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789+/";

    fn encode_sextet(v: u8) -> char {
        char::from(Self::CHARSET[usize::from(v)])
    }

    fn decode_sextet(b: u8) -> Result<u8, EvalError> {
        match b {
            b'A'..=b'Z' => Ok(b - b'A'),
            b'a'..=b'z' => Ok(b - b'a' + 26),
            b'+' => Ok(62),
            b'/' => Ok(63),
            _ => Err(EvalError::InvalidBase64Symbol(char::from(b))),
        }
    }
}

impl Format for Base64Format {
    // Support for PostgreSQL-style (which is really MIME-style) Base64 encoding
    // was, frustratingly, removed from Rust's `base64` crate. So we roll our
    // own Base64 encoder and decoder here.

    fn encode(&self, bytes: &[u8]) -> String {
        // Process input in chunks of three octets. Each chunk is converted to
        // four sextets. Each sextet is encoded as a printable ASCII character
        // via `CHARSET`.
        //
        // When the input length is not divisible by three, the last chunk is
        // partial. Sextets that are entirely determined by missing octets are
        // encoded as `=`. Sextets that are partially determined by a missing
        // octect assume the octet was zero.
        //
        // Line breaks are emitted after every 76 characters.

        let mut buf = String::new();
        for chunk in bytes.chunks(3) {
            match chunk {
                [o1, o2, o3] => {
                    let s1 = (o1 & 0b11111100) >> 2;
                    let s2 = (o1 & 0b00000011) << 4 | (o2 & 0b11110000) >> 4;
                    let s3 = (o2 & 0b00001111) << 2 | (o3 & 0b11000000) >> 6;
                    let s4 = o3 & 0b00111111;
                    buf.push(Self::encode_sextet(s1));
                    buf.push(Self::encode_sextet(s2));
                    buf.push(Self::encode_sextet(s3));
                    buf.push(Self::encode_sextet(s4));
                }
                [o1, o2] => {
                    let s1 = (o1 & 0b11111100) >> 2;
                    let s2 = (o1 & 0b00000011) << 4 | (o2 & 0b11110000) >> 4;
                    let s3 = (o2 & 0b00001111) << 2;
                    buf.push(Self::encode_sextet(s1));
                    buf.push(Self::encode_sextet(s2));
                    buf.push(Self::encode_sextet(s3));
                    buf.push('=');
                }
                [o1] => {
                    let s1 = (o1 & 0b11111100) >> 2;
                    let s2 = (o1 & 0b00000011) << 4;
                    buf.push(Self::encode_sextet(s1));
                    buf.push(Self::encode_sextet(s2));
                    buf.push('=');
                    buf.push('=');
                }
                _ => unreachable!(),
            }
            if buf.len() % 77 == 76 {
                buf.push('\n');
            }
        }
        buf
    }

    fn decode(&self, s: &str) -> Result<Vec<u8>, EvalError> {
        // Process input in chunks of four bytes, after filtering out any bytes
        // that represent whitespace. Each byte is decoded into a sextet
        // according to the reverse charset mapping maintained in
        // `Self::decode_sextet`. The four sextets are converted to three octets
        // and emitted.
        //
        // When the last character in a chunk is `=` or the last two characters
        // in a chunk are both `=`, the chunk is missing its last one or two
        // sextets, respectively. Octets that are entirely determined by missing
        // sextets are elided. Octets that are partially determined by a missing
        // sextet assume the sextet was zero.
        //
        // It is an error for a `=` character to appear in another position in
        // a chunk. It is also an error if a chunk is incomplete.

        let mut buf = vec![];
        let mut bytes = s
            .as_bytes()
            .iter()
            .copied()
            .filter(|ch| !matches!(ch, b' ' | b'\t' | b'\n' | b'\r'));
        loop {
            match (bytes.next(), bytes.next(), bytes.next(), bytes.next()) {
                (Some(c1), Some(c2), Some(b'='), Some(b'=')) => {
                    let s1 = Self::decode_sextet(c1)?;
                    let s2 = Self::decode_sextet(c2)?;
                    buf.push(s1 << 2 | (s2 & 0b110000) >> 4);
                }
                (Some(c1), Some(c2), Some(c3), Some(b'=')) => {
                    let s1 = Self::decode_sextet(c1)?;
                    let s2 = Self::decode_sextet(c2)?;
                    let s3 = Self::decode_sextet(c3)?;
                    buf.push(s1 << 2 | (s2 & 0b110000) >> 4);
                    buf.push((s2 & 0b001111) << 4 | (s3 & 0b111100) >> 2);
                }
                (Some(b'='), _, _, _) | (_, Some(b'='), _, _) | (_, _, Some(b'='), _) => {
                    return Err(EvalError::InvalidBase64Equals)
                }
                (Some(c1), Some(c2), Some(c3), Some(c4)) => {
                    let s1 = Self::decode_sextet(c1)?;
                    let s2 = Self::decode_sextet(c2)?;
                    let s3 = Self::decode_sextet(c3)?;
                    let s4 = Self::decode_sextet(c4)?;
                    buf.push(s1 << 2 | (s2 & 0b110000) >> 4);
                    buf.push((s2 & 0b001111) << 4 | (s3 & 0b111100) >> 2);
                    buf.push((s3 & 0b000011) << 6 | s4);
                }
                (None, None, None, None) => return Ok(buf),
                _ => return Err(EvalError::InvalidBase64EndSequence),
            }
        }
    }
}

struct EscapeFormat;

impl Format for EscapeFormat {
    fn encode(&self, bytes: &[u8]) -> String {
        let mut buf = String::new();
        for b in bytes {
            match b {
                b'\0' | (b'\x80'..=b'\xff') => {
                    buf.push('\\');
                    write!(&mut buf, "{:03o}", b);
                }
                b'\\' => buf.push_str("\\\\"),
                _ => buf.push(char::from(*b)),
            }
        }
        buf
    }

    fn decode(&self, s: &str) -> Result<Vec<u8>, EvalError> {
        Ok(strconv::parse_bytes_traditional(s)?)
    }
}

struct HexFormat;

impl Format for HexFormat {
    fn encode(&self, bytes: &[u8]) -> String {
        hex::encode(bytes)
    }

    fn decode(&self, s: &str) -> Result<Vec<u8>, EvalError> {
        // Can't use `hex::decode` here, as it doesn't tolerate whitespace
        // between encoded bytes.
        Ok(strconv::parse_bytes_hex(s)?)
    }
}

pub fn lookup_format(s: &str) -> Result<&'static dyn Format, EvalError> {
    let s = UncasedStr::new(s);
    if s == "base64" {
        Ok(&Base64Format)
    } else if s == "escape" {
        Ok(&EscapeFormat)
    } else if s == "hex" {
        Ok(&HexFormat)
    } else {
        Err(EvalError::InvalidEncodingName(s.as_str().into()))
    }
}
