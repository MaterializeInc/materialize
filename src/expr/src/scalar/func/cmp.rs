// Copyright Materialize, Inc. All rights reserved.
//
// Use of this software is governed by the Business Source License
// included in the LICENSE file.
//
// As of the Change Date specified in that file, in accordance with
// the Business Source License, use of this software will be governed
// by the Apache License, Version 2.0.

//! Comparison functions.

use repr::{Datum, ScalarType};

use crate::scalar::func::{FuncProps, Nulls, OutputType};
use crate::scalar::EvalError;

pub const CMP_PROPS: FuncProps = FuncProps {
    can_error: false,
    preserves_uniqueness: false,
    nulls: Nulls::Sometimes {
        propagates_nulls: true,
        introduces_nulls: false,
    },
    output_type: OutputType::Fixed(ScalarType::Bool),
};

pub fn eq<'a>(a: Datum<'a>, b: Datum<'a>) -> Result<Datum<'a>, EvalError> {
    Ok(Datum::from(a == b))
}

pub fn not_eq<'a>(a: Datum<'a>, b: Datum<'a>) -> Result<Datum<'a>, EvalError> {
    Ok(Datum::from(a != b))
}

pub fn lt<'a>(a: Datum<'a>, b: Datum<'a>) -> Result<Datum<'a>, EvalError> {
    Ok(Datum::from(a < b))
}

pub fn lte<'a>(a: Datum<'a>, b: Datum<'a>) -> Result<Datum<'a>, EvalError> {
    Ok(Datum::from(a <= b))
}

pub fn gt<'a>(a: Datum<'a>, b: Datum<'a>) -> Result<Datum<'a>, EvalError> {
    Ok(Datum::from(a > b))
}

pub fn gte<'a>(a: Datum<'a>, b: Datum<'a>) -> Result<Datum<'a>, EvalError> {
    Ok(Datum::from(a >= b))
}
