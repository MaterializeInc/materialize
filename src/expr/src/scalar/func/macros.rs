// Copyright Materialize, Inc. and contributors. All rights reserved.
//
// Use of this software is governed by the Business Source License
// included in the LICENSE file.
//
// As of the Change Date specified in that file, in accordance with
// the Business Source License, use of this software will be governed
// by the Apache License, Version 2.0.

macro_rules! sqlfunc {
    // Expand the function name into an attribute if it was omitted
    (
        fn $fn_name:ident $($tail:tt)*
    ) => {
        sqlfunc!(
            #[sqlname = stringify!($fn_name)]
            fn $fn_name $($tail)*
        );
    };

    // Add the uniqueness attribute if it was omitted
    (
        #[sqlname = $name:expr]
        fn $fn_name:ident $($tail:tt)*
    ) => {
        sqlfunc!(
            #[sqlname = $name]
            #[preserves_uniqueness = false]
            fn $fn_name $($tail)*
        );
    };

    // Add lifetime parameter if it was omitted
    (
        #[sqlname = $name:expr]
        #[preserves_uniqueness = $preserves_uniqueness:expr]
        fn $fn_name:ident ($($params:tt)*) $($tail:tt)*
    ) => {
        sqlfunc!(
            #[sqlname = $name]
            #[preserves_uniqueness = $preserves_uniqueness]
            fn $fn_name<'a>($($params)*) $($tail)*
        );
    };

    // Normalize mut arguments to non-mut ones
    (
        #[sqlname = $name:expr]
        #[preserves_uniqueness = $preserves_uniqueness:expr]
        fn $fn_name:ident<$lt:lifetime>(mut $param_name:ident: $input_ty:ty) -> $output_ty:ty
            $body:block
    ) => {
        sqlfunc!(
            #[sqlname = $name]
            #[preserves_uniqueness = $preserves_uniqueness]
            fn $fn_name<$lt>($param_name: $input_ty) -> $output_ty {
                let mut $param_name = $param_name;
                $body
            }
        );
    };

    (
        #[sqlname = $name:expr]
        #[preserves_uniqueness = $preserves_uniqueness:expr]
        fn $fn_name:ident<$lt:lifetime>($param_name:ident: $input_ty:ty) -> $output_ty:ty
            $body:block
    ) => {
        paste::paste! {
            #[derive(proptest_derive::Arbitrary, Ord, PartialOrd, Clone, Debug, Eq, PartialEq, serde::Serialize, serde::Deserialize, Hash, mz_lowertest::MzReflect)]
            pub struct [<$fn_name:camel>];

            impl<'a> crate::func::EagerUnaryFunc<'a> for [<$fn_name:camel>] {
                type Input = $input_ty;
                type Output = $output_ty;

                fn call(&self, a: Self::Input) -> Self::Output {
                    $fn_name(a)
                }

                fn output_type(&self, input_type: mz_repr::ColumnType) -> mz_repr::ColumnType {
                    use mz_repr::AsColumnType;
                    let output = Self::Output::as_column_type();
                    let propagates_nulls = crate::func::EagerUnaryFunc::propagates_nulls(self);
                    let nullable = output.nullable;
                    // The output is nullable if it is nullable by itself or the input is nullable
                    // and this function propagates nulls
                    output.nullable(nullable || (propagates_nulls && input_type.nullable))
                }

                fn preserves_uniqueness(&self) -> bool {
                    $preserves_uniqueness
                }
            }

            impl std::fmt::Display for [<$fn_name:camel>] {
                fn fmt(&self, f: &mut std::fmt::Formatter) -> std::fmt::Result {
                    f.write_str($name)
                }
            }

            pub fn $fn_name<$lt>($param_name: $input_ty) -> $output_ty {
                $body
            }
        }
    };
}

#[cfg(test)]
mod test {
    use crate::scalar::func::LazyUnaryFunc;
    use crate::EvalError;
    use mz_repr::ScalarType;

    sqlfunc!(
        #[sqlname = "INFALLIBLE"]
        fn infallible1(a: f32) -> f32 {
            a
        }
    );

    sqlfunc!(
        fn infallible2(a: Option<f32>) -> f32 {
            a.unwrap_or_default()
        }
    );

    sqlfunc!(
        fn infallible3(a: f32) -> Option<f32> {
            Some(a)
        }
    );

    #[test]
    fn elision_rules_infallible() {
        assert_eq!(format!("{}", Infallible1), "INFALLIBLE");
        assert!(Infallible1.propagates_nulls());
        assert!(!Infallible1.introduces_nulls());

        assert!(!Infallible2.propagates_nulls());
        assert!(!Infallible2.introduces_nulls());

        assert!(Infallible3.propagates_nulls());
        assert!(Infallible3.introduces_nulls());
    }

    #[test]
    fn output_types_infallible() {
        assert_eq!(
            Infallible1.output_type(ScalarType::Float32.nullable(true)),
            ScalarType::Float32.nullable(true)
        );
        assert_eq!(
            Infallible1.output_type(ScalarType::Float32.nullable(false)),
            ScalarType::Float32.nullable(false)
        );

        assert_eq!(
            Infallible2.output_type(ScalarType::Float32.nullable(true)),
            ScalarType::Float32.nullable(false)
        );
        assert_eq!(
            Infallible2.output_type(ScalarType::Float32.nullable(false)),
            ScalarType::Float32.nullable(false)
        );

        assert_eq!(
            Infallible3.output_type(ScalarType::Float32.nullable(true)),
            ScalarType::Float32.nullable(true)
        );
        assert_eq!(
            Infallible3.output_type(ScalarType::Float32.nullable(false)),
            ScalarType::Float32.nullable(true)
        );
    }

    sqlfunc!(
        fn fallible1(a: f32) -> Result<f32, EvalError> {
            Ok(a)
        }
    );

    sqlfunc!(
        fn fallible2(a: Option<f32>) -> Result<f32, EvalError> {
            Ok(a.unwrap_or_default())
        }
    );

    sqlfunc!(
        fn fallible3(a: f32) -> Result<Option<f32>, EvalError> {
            Ok(Some(a))
        }
    );

    #[test]
    fn elision_rules_fallible() {
        assert!(Fallible1.propagates_nulls());
        assert!(!Fallible1.introduces_nulls());

        assert!(!Fallible2.propagates_nulls());
        assert!(!Fallible2.introduces_nulls());

        assert!(Fallible3.propagates_nulls());
        assert!(Fallible3.introduces_nulls());
    }

    #[test]
    fn output_types_fallible() {
        assert_eq!(
            Fallible1.output_type(ScalarType::Float32.nullable(true)),
            ScalarType::Float32.nullable(true)
        );
        assert_eq!(
            Fallible1.output_type(ScalarType::Float32.nullable(false)),
            ScalarType::Float32.nullable(false)
        );

        assert_eq!(
            Fallible2.output_type(ScalarType::Float32.nullable(true)),
            ScalarType::Float32.nullable(false)
        );
        assert_eq!(
            Fallible2.output_type(ScalarType::Float32.nullable(false)),
            ScalarType::Float32.nullable(false)
        );

        assert_eq!(
            Fallible3.output_type(ScalarType::Float32.nullable(true)),
            ScalarType::Float32.nullable(true)
        );
        assert_eq!(
            Fallible3.output_type(ScalarType::Float32.nullable(false)),
            ScalarType::Float32.nullable(true)
        );
    }
}

/// Temporary macro that generates the equivalent of what enum_dispatch will do in the future. We
/// need this manual macro implementation to delegate to the previous manual implementation for
/// variants that use the old definitions.
///
/// Once everything is handled by this macro we can remove it and replace it with `enum_dispatch`
macro_rules! derive_unary {
    ($($name:ident),*) => {
        impl UnaryFunc {
            pub fn eval<'a>(
                &'a self,
                datums: &[Datum<'a>],
                temp_storage: &'a RowArena,
                a: &'a MirScalarExpr,
            ) -> Result<Datum<'a>, EvalError> {
                match self {
                    $(Self::$name(f) => f.eval(datums, temp_storage, a),)*
                    _ => self.eval_manual(datums, temp_storage, a),
                }
            }

            pub fn output_type(&self, input_type: ColumnType) -> ColumnType {
                match self {
                    $(Self::$name(f) => LazyUnaryFunc::output_type(f, input_type),)*
                    _ => self.output_type_manual(input_type),
                }
            }
            pub fn propagates_nulls(&self) -> bool {
                match self {
                    $(Self::$name(f) => LazyUnaryFunc::propagates_nulls(f),)*
                    _ => self.propagates_nulls_manual(),
                }
            }
            pub fn introduces_nulls(&self) -> bool {
                match self {
                    $(Self::$name(f) => LazyUnaryFunc::introduces_nulls(f),)*
                    _ => self.introduces_nulls_manual(),
                }
            }
            pub fn preserves_uniqueness(&self) -> bool {
                match self {
                    $(Self::$name(f) => LazyUnaryFunc::preserves_uniqueness(f),)*
                    _ => self.preserves_uniqueness_manual(),
                }
            }
        }

        impl fmt::Display for UnaryFunc {
            fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
                match self {
                    $(Self::$name(func) => func.fmt(f),)*
                    _ => self.fmt_manual(f),
                }
            }
        }

        $(
            impl From<$name> for crate::UnaryFunc {
                fn from(variant: $name) -> Self {
                    Self::$name(variant)
                }
            }
        )*
    }
}
