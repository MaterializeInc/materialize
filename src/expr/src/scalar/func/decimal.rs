// Copyright Materialize, Inc. All rights reserved.
//
// Use of this software is governed by the Business Source License
// included in the LICENSE file.
//
// As of the Change Date specified in that file, in accordance with
// the Business Source License, use of this software will be governed
// by the Apache License, Version 2.0.

//! Decimal functions.

use std::cmp::Ordering;

use ore::collections::CollectionExt;
use ore::result::ResultExt;
use repr::adt::decimal::MAX_DECIMAL_PRECISION;
use repr::{strconv, Datum, RowArena, ScalarType};

use crate::scalar::func::{float, FuncProps, Nulls, OutputType};
use crate::scalar::EvalError;

pub const CAST_DECIMAL_TO_STRING_PROPS: FuncProps = FuncProps {
    can_error: false,
    preserves_uniqueness: false,
    nulls: Nulls::Sometimes {
        propagates_nulls: true,
        introduces_nulls: false,
    },
    output_type: OutputType::Fixed(ScalarType::String),
};

pub fn cast_decimal_to_string<'a>(
    a: Datum<'a>,
    scale: u8,
    temp_storage: &'a RowArena,
) -> Result<Datum<'a>, EvalError> {
    let mut buf = String::new();
    strconv::format_decimal(&mut buf, &a.unwrap_decimal().with_scale(scale));
    Ok(Datum::String(temp_storage.push_string(buf)))
}

pub fn cast_string_to_decimal_props(scale: u8) -> FuncProps {
    FuncProps {
        can_error: true,
        preserves_uniqueness: false,
        nulls: Nulls::Sometimes {
            propagates_nulls: true,
            introduces_nulls: false,
        },
        output_type: OutputType::Fixed(ScalarType::Decimal(MAX_DECIMAL_PRECISION, scale)),
    }
}

pub fn cast_string_to_decimal(a: Datum, scale: u8) -> Result<Datum, EvalError> {
    strconv::parse_decimal(a.unwrap_str())
        .map(|d| {
            Datum::from(match d.scale().cmp(&scale) {
                Ordering::Less => d.significand() * 10_i128.pow(u32::from(scale - d.scale())),
                Ordering::Equal => d.significand(),
                Ordering::Greater => d.significand() / 10_i128.pow(u32::from(d.scale() - scale)),
            })
        })
        .err_into()
}

pub const CAST_DECIMAL_TO_INT32_PROPS: FuncProps = FuncProps {
    can_error: false,
    preserves_uniqueness: false,
    nulls: Nulls::Sometimes {
        propagates_nulls: true,
        introduces_nulls: false,
    },
    output_type: OutputType::Fixed(ScalarType::Int32),
};

pub fn cast_decimal_to_int32(a: Datum) -> Result<Datum, EvalError> {
    Ok(Datum::from(a.unwrap_decimal().as_i128() as i32))
}

pub const CAST_DECIMAL_TO_INT64_PROPS: FuncProps = FuncProps {
    can_error: false,
    preserves_uniqueness: false,
    nulls: Nulls::Sometimes {
        propagates_nulls: true,
        introduces_nulls: false,
    },
    output_type: OutputType::Fixed(ScalarType::Int64),
};

pub fn cast_decimal_to_int64(a: Datum) -> Result<Datum, EvalError> {
    Ok(Datum::from(a.unwrap_decimal().as_i128() as i64))
}

pub const CAST_DECIMAL_TO_FLOAT32_PROPS: FuncProps = FuncProps {
    can_error: false,
    preserves_uniqueness: false,
    nulls: Nulls::Sometimes {
        propagates_nulls: true,
        introduces_nulls: false,
    },
    output_type: OutputType::Fixed(ScalarType::Float32),
};

pub fn cast_decimal_to_float32(a: Datum) -> Result<Datum, EvalError> {
    Ok(Datum::from(a.unwrap_decimal().as_i128() as f32))
}

pub const CAST_DECIMAL_TO_FLOAT64_PROPS: FuncProps = FuncProps {
    can_error: false,
    preserves_uniqueness: false,
    nulls: Nulls::Sometimes {
        propagates_nulls: true,
        introduces_nulls: false,
    },
    output_type: OutputType::Fixed(ScalarType::Float64),
};

pub fn cast_decimal_to_float64(a: Datum) -> Result<Datum, EvalError> {
    Ok(Datum::from(a.unwrap_decimal().as_i128() as f64))
}

pub const DECIMAL_MATH_PROPS: FuncProps = FuncProps {
    can_error: true,
    preserves_uniqueness: false,
    nulls: Nulls::Sometimes {
        propagates_nulls: true,
        introduces_nulls: false,
    },
    output_type: OutputType::MatchesInput,
};

pub fn abs_decimal(a: Datum) -> Result<Datum, EvalError> {
    Ok(Datum::from(a.unwrap_decimal().abs()))
}

pub fn neg_decimal(a: Datum) -> Result<Datum, EvalError> {
    Ok(Datum::from(-a.unwrap_decimal()))
}

pub fn add_decimal<'a>(a: Datum<'a>, b: Datum<'a>) -> Result<Datum<'a>, EvalError> {
    Ok(Datum::from(a.unwrap_decimal() + b.unwrap_decimal()))
}

pub fn sub_decimal<'a>(a: Datum<'a>, b: Datum<'a>) -> Result<Datum<'a>, EvalError> {
    Ok(Datum::from(a.unwrap_decimal() - b.unwrap_decimal()))
}

pub fn mod_decimal<'a>(a: Datum<'a>, b: Datum<'a>) -> Result<Datum<'a>, EvalError> {
    let b = b.unwrap_decimal();
    if b == 0 {
        Err(EvalError::DivisionByZero)
    } else {
        Ok(Datum::from(a.unwrap_decimal() % b))
    }
}

pub const MUL_DECIMAL_PROPS: FuncProps = FuncProps {
    can_error: true,
    preserves_uniqueness: false,
    nulls: Nulls::Sometimes {
        propagates_nulls: true,
        introduces_nulls: false,
    },
    output_type: OutputType::Computed(|input_types| {
        assert_eq!(input_types.len(), 2);
        let (_, s1) = input_types[0].scalar_type.unwrap_decimal_parts();
        let (_, s2) = input_types[1].scalar_type.unwrap_decimal_parts();
        ScalarType::Decimal(MAX_DECIMAL_PRECISION, s1 + s2)
    }),
};

pub fn mul_decimal<'a>(a: Datum<'a>, b: Datum<'a>) -> Result<Datum<'a>, EvalError> {
    Ok(Datum::from(a.unwrap_decimal() * b.unwrap_decimal()))
}

pub const DIV_DECIMAL_PROPS: FuncProps = FuncProps {
    can_error: true,
    preserves_uniqueness: false,
    nulls: Nulls::Sometimes {
        propagates_nulls: true,
        introduces_nulls: false,
    },
    output_type: OutputType::Computed(|input_types| {
        assert_eq!(input_types.len(), 2);
        let (_, s1) = input_types[0].scalar_type.unwrap_decimal_parts();
        let (_, s2) = input_types[1].scalar_type.unwrap_decimal_parts();
        ScalarType::Decimal(MAX_DECIMAL_PRECISION, s1 - s2)
    }),
};

pub fn div_decimal<'a>(a: Datum<'a>, b: Datum<'a>) -> Result<Datum<'a>, EvalError> {
    let b = b.unwrap_decimal();
    if b == 0 {
        Err(EvalError::DivisionByZero)
    } else {
        Ok(Datum::from(a.unwrap_decimal() / b))
    }
}

pub fn ceil_decimal(a: Datum, scale: u8) -> Result<Datum, EvalError> {
    let decimal = a.unwrap_decimal();
    Ok(Datum::from(decimal.with_scale(scale).ceil().significand()))
}

pub fn floor_decimal(a: Datum, scale: u8) -> Result<Datum, EvalError> {
    let decimal = a.unwrap_decimal();
    Ok(Datum::from(decimal.with_scale(scale).floor().significand()))
}

pub const ROUND_DECIMAL_PROPS: FuncProps = FuncProps {
    can_error: false,
    preserves_uniqueness: false,
    nulls: Nulls::Sometimes {
        propagates_nulls: true,
        introduces_nulls: false,
    },
    output_type: OutputType::Computed(|input_types| input_types.into_first().scalar_type),
};

pub fn round_decimal_unary(a: Datum, a_scale: u8) -> Result<Datum, EvalError> {
    round_decimal_binary(a, Datum::Int64(0), a_scale)
}

pub fn round_decimal_binary<'a>(
    a: Datum<'a>,
    b: Datum<'a>,
    a_scale: u8,
) -> Result<Datum<'a>, EvalError> {
    let round_to = b.unwrap_int64();
    let decimal = a.unwrap_decimal().with_scale(a_scale);
    Ok(Datum::from(decimal.round(round_to).significand()))
}

pub const SQRT_DECIMAL_PROPS: FuncProps = FuncProps {
    can_error: true,
    preserves_uniqueness: false,
    nulls: Nulls::Sometimes {
        propagates_nulls: true,
        introduces_nulls: false,
    },
    output_type: OutputType::MatchesInput,
};

pub fn sqrt_decimal(a: Datum, scale: u8) -> Result<Datum, EvalError> {
    let d = a.unwrap_decimal();
    if d.as_i128() < 0 {
        return Err(EvalError::NegSqrt);
    }
    let d_f64 = cast_decimal_to_float64(a)?;
    let d_scaled = d_f64.unwrap_float64() / 10_f64.powi(i32::from(scale));
    float::cast_float64_to_decimal(Datum::from(d_scaled.sqrt()), scale)
}
