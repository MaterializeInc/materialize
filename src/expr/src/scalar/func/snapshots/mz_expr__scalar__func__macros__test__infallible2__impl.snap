---
source: src/expr/src/scalar/func/macros.rs
expression: info
---
Info {
    output_type_nullable: SqlColumnType {
        scalar_type: Float32,
        nullable: false,
    },
    output_type_nonnullable: SqlColumnType {
        scalar_type: Float32,
        nullable: false,
    },
    preserves_uniqueness: false,
    inverse: None,
    is_monotone: false,
    propagates_nulls: false,
    introduces_nulls: false,
    could_error: false,
}
